types/src/auction.rs

```diff
@@ -263,18 +266,34 @@ pub trait Auction:
             return Err(Error::InvalidCaller);
         }
 
-        let mut burned_amount = detail::quash_bid(self, &validator_public_keys)?;
+        let mut burned_amount: U512 = U512::zero();
 
-        let mut unbonding_purses: UnbondingPurses = detail::get_unbonding_purses(self)?;
+        let mut bids = detail::get_bids(self)?;
+        let mut modified_validators = 0usize;
 
+        let mut unbonding_purses: UnbondingPurses = detail::get_unbonding_purses(self)?;
         let mut unbonding_purses_modified = false;
         for validator_public_key in validator_public_keys {
+            // Clean up bids for given validator and save up assigned delegators
+            let orphaned_delegators = match bids.remove(&validator_public_key) {
+                Some(bid) => {
+                    burned_amount += *bid.staked_amount();
+                    modified_validators += 1;
+
+                    bid.delegators().keys().copied().collect()
+                }
+                None => BTreeSet::new(),
+            };
+
+            // Clean up unbonding entry for given validator
             if let Some(unbonding_list) = unbonding_purses.get_mut(&validator_public_key) {
                 let initial_length = unbonding_list.len();
 
                 unbonding_list.retain(|unbonding_purse| {
-                    // Only entries created by non-validators are retained
-                    let should_retain = !unbonding_purse.is_validator();
+                    // Only entries created by validator entries and delegators that were attached
+                    // to quashed bids of validator.
```

Comment:
```suggestion
                    // Retain all entries where this validator is delegating to another validator.
```

(if my assumptions in the other comment are correct)